### ISSUE TITLE
Retina mask is shifted

### ISSUE THREAD

**Issue by stephencheung16** [ultralytics/yolov5#11908]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### Bug

Install Yolov5 & Roboflow.  Download weight of pre-trained Yolov5x segmentation model.
![p1](https://github.com/ultralytics/yolov5/assets/23723498/3b7120a7-cbbf-45ac-900a-7d6d4ca4ca88)

Run prediction on skis with --retina-masks option.  Retina mask is shifted to bottom-right.
![p2](https://github.com/ultralytics/yolov5/assets/23723498/5f1fc663-74e2-494e-a249-7d3f5ce42667)

Run prediction on skis without --retina-masks option.  The mask is not shifted.
![p3](https://github.com/ultralytics/yolov5/assets/23723498/94d65b4a-9083-43d9-b575-7eda9fb4ad1b)

Would you please help?

### Environment

- Kaggle notebook

### Minimal Reproducible Example

# Install Yolov5
!git clone https://github.com/ultralytics/yolov5  # clone
%cd yolov5
%pip install -qr requirements.txt  # install

import torch
import utils
display = utils.notebook_init()  # checks

# Test if retina mask is shifted on skis
image_path = '/kaggle/input/test-images/skis_600x396.jpg'
!python segment/predict.py --img 320 --weights yolov5x-seg.pt --source {image_path} \
    --save-txt --save-conf --device cpu \
    --retina-masks

display.Image(filename=f'runs/predict-seg/exp/skis_600x396.jpg', width=600)

# Test if mask is shifted if --retina-mask option is deleted
image_path = '/kaggle/input/test-images/skis_600x396.jpg'
!python segment/predict.py --img 320 --weights yolov5x-seg.pt --source {image_path} \
    --save-txt --save-conf --device cpu

display.Image(filename=f'runs/predict-seg/exp2/skis_600x396.jpg', width=600)

# The test image
![skis_600x396](https://github.com/ultralytics/yolov5/assets/23723498/c69f8a7e-33f6-417b-93e4-91adc5bce374)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@stephencheung16 hi there! Thank you for bringing this issue to our attention. I understand that the retina mask is shifted when using the `--retina-masks` option in YOLOv5 segmentation prediction.

To better assist you, could you please confirm if you have followed the steps in the [YOLOv5 tutorial](https://docs.ultralytics.com/yolov5/) correctly, specifically the instructions for setting up the environment and installing the requirements? 

Additionally, it would be helpful if you could provide the following information:
- Versions of Python, PyTorch, and the YOLOv5 codebase you are using.
- Any modifications made to the YOLOv5 codebase or custom configurations.

Please ensure that you are using the latest version of YOLOv5 and its dependencies. If the issue persists, please provide a **minimum reproducible example** that we can use to debug the problem.

Thank you for your collaboration, and we look forward to resolving this issue with you.

**Comment by stephencheung16** [comment]
@glenn-jocher

Hi Glenn,

I follow the steps in the Roboflow tutorial 'How to Train YOLOv5 Instance Segmentation on a Custom Dataset' for setting up the environment and installing the requirements

In Kaggle, the Python version is 3.10.12. Pytorch version is 2.0.0.  YOLOv5 is v7.0-196-gacdf73b.

I did not modify any of the YOLOv5 codebase or custom configurations.

The 'minimum reproducible example' is attached above, in my first message.  Both code and test image are attached.

Just modify the path to the test image.  The code can be run.  There are only 12 lines.

Please try the 'minimum reproducible example'.  Thanks.

**Comment by glenn-jocher** [comment]
Hi @stephencheung16!

Thank you for providing the details and the minimum reproducible example. I apologize for any confusion caused.

I understand that you followed the steps in the Roboflow tutorial for setting up the environment and installing the requirements. I also see that you are using Python version 3.10.12, PyTorch version 2.0.0, and YOLOv5 version v7.0-196-gacdf73b.

I will try the minimum reproducible example you provided with the test image you attached. I will investigate the issue and get back to you with a solution or further information as soon as possible.

Thank you for your patience, and I appreciate your cooperation in resolving this issue.

Best regards,
Glenn

**Comment by stephencheung16** [comment]
Hi @glenn-jocher,

By making the skis thinner, the shift is more obvious.

With --retina-masks,
![skis_600x396_further_narrowed](https://github.com/ultralytics/yolov5/assets/23723498/71e5712e-47d4-4d24-af49-14bb0cdf6c39)

Without --retina-masks,
![skis_600x396_further_narrowed](https://github.com/ultralytics/yolov5/assets/23723498/1d51b64e-18cb-45ea-9201-f98d5f6c44e7)

Attached is the test image,
![skis_600x396_further_narrowed](https://github.com/ultralytics/yolov5/assets/23723498/7ab42c48-34ca-4a11-990a-5f9d2fe0e039)

Would you please help?  Thanks.

**Comment by glenn-jocher** [comment]
Hi @stephencheung16,

Thank you for providing the details and additional images.

I understand that when using the `--retina-masks` option in YOLOv5 for segmentation predictions, the shift becomes more obvious, particularly when the skis are thinner. In comparison, without the `--retina-masks` option, the shift is less apparent.

To investigate and address this issue, I will analyze the provided test image and run additional tests with the `--retina-masks` option enabled.

I appreciate your patience, and I will get back to you with further information or a solution as soon as possible.

Thank you for your understanding.

Best regards,
Glenn

**Comment by Boatsure** [comment]
Interesting, now is 2024 and I use yolov8.1-seg to do instance segmentation inference, the visualization masks will be shifted to bottom-right with retina_masks True and imgsz **not** 640. I mean the masks won't be shifted when imgsz=640, with retina_masks on or not.

**Comment by glenn-jocher** [comment]
Hi there!

Thank you for sharing your findings with YOLOv8.1-seg regarding the mask shifting issue related to the `imgsz` parameter and `retina_masks` option. It seems like resizing the images to sizes other than 640 pixels might be affecting the alignment of the masks.

A potential workaround could be to ensure all image inputs are resized to 640 pixels before running segmentation, as this size seems to prevent the shifting issue. You might also check for any discrepancies in the resizing or coordinate transformation logic in the code if adjusting `imgsz` is not feasible for your application.

Your input is valuable, and we'll look into this for potential improvements in future updates! 🌟
